### ISSUE TITLE
Descriptor leaks in multithreaded environment

### ISSUE THREAD

**Issue by szjur** [saltstack/salt#36374]
### Description of Issue/Question

I ran into an issue similar to https://github.com/saltstack/salt/issues/20639. In multithreaded environment LocalClient leaks descriptor to pipes, sockets and that anon_inode:[eventfd] mentioned in #20639. But there is more to it. It seems that using cmd_async() seems to trigger the leaks. 

In my use case I want to listen for events from a central thread and submit async jobs from a bunch of others. That seems to result in this massive proliferation of open handles - I think it's 4 per thread created. Those handles stay open after the thread ends.

Please see the code below.
### Setup

Any simple 2016.3.2 setup.
### Steps to Reproduce Issue

Run the code below and watch /proc/pid/fd growing.

```
import os, threading, time
import salt.client

class TestThread(threading.Thread):
        def run(self):
                self.salt_local_client = salt.client.LocalClient()
                print self.salt_local_client.cmd_async(tgt = ["minion.name "], fun = "cmd.run", arg = ["uname -a"], expr_form="list")
                time.sleep(8)
                del self.salt_local_client

print "PID: %s" % os.getpid()
for _ in range(50):
        test_thread = TestThread()
        test_thread.run()
        time.sleep(10)
```

Now change the run() function method to use cmd():

```
class TestThread(threading.Thread):
        def run(self):
                self.salt_local_client = salt.client.LocalClient()
                print self.salt_local_client.cmd(tgt = ["minion.name "], fun = "cmd.run", arg = ["uname -a"], expr_form="list")
                del self.salt_local_client
```

and see /proc/pid/fd stabilise at around 10-15 descriptors.

I also added a debug to SaltEvent.destroy() method and in the cmd_async() variant it doesn't seem to be called. I tried calling it manually with self.salt_local_client.event.destroy() just for testing and it helps in the above example. But when that trick is applied to my use case then it somehow messes up the thread which listens for all events (without any filter).
### Versions Report

2016.3.2

**Comment by Ch3LL** [comment]
@szjur i'm having a hard time replicating this. I'm only seeing the FDs going around 5-15 on the first example you provided:

```
 ch3ll@thecakeisalie  ~/git/salt  ➦ 721e6dc  docker exec salt-cent7 /bin/bash -c "cd /proc/1493/fd/; ls -l|wc -l"
9
```

I will provide a docker container so maybe you can see what is wrong with my example:

1. `docker run -it -v /home/ch3ll/git/salt/:/testing/ ch3ll/issues:36374 /bin/bash` (/home/ch3ll/git/salt being a local checkout of salt on tag version 2016.3.2)
2. `salt-master -d`
3. `salt-minion -d`
4. `python /tmp/test.py`
5. `docker exec <image> /bin/bash -c "cd /proc/<proc>/fd/; ls -l|wc -l"`

Maybe you can point out a misunderstanding on my point of the test case i am using above. thanks

**Comment by szjur** [comment]
You are checking the right process, right? I'm not really sure how that behaves in docker. I actually forgot that I also have syndics in my environment and that affects how the client works (there is "if self.opts['order_masters']" etc here and there in https://github.com/saltstack/salt/blob/2016.3/salt/client/__init__.py).
What I found out today after analysing the client code was that if I call _clean_up_subscriptions(jid) after getting a jid from cmd_async() that will stop the leaks. Looking at the code I cannot explain it why though. I believe the client calls self.event.subscribe() in the cmd_async() workflow and apparently for some reason active event subscriptions prevent garbage collector to clean up the client properly. It may depend on Python version too - unfortunately I currently need to use an ancient 2.6 version from RHEL 6.
Maybe your developers can somehow make sense of that information.

**Comment by Ch3LL** [comment]
@szjur yes i'm using the PID thats printed out by the code you provided. I will ping someoen more knowledgeable on this one to take a look at this issue. Thanks 

ping @cachedout

**Comment by cachedout** [comment]
We just did a big rework of the client here: https://github.com/saltstack/salt/pull/36720/files

Any chance you could re-test this against the head of the carbon branch?

**Comment by szjur** [comment]
Sure, but I'm afraid that can take a while as I'm now doing a project already based on 2016.3 which if fails it may abruptly end my relationship with SaltStack ;-) The workaround mentioned above seems to stop descriptor leak, although in some abnormal situations I think it can still leak a bit (but the impact is much less severe). If Ch3LL couldn't reproduce it, maybe it's RHEL 6 / Python 2.6 specific.

**Comment by tsaridas** [comment]
@cachedout , @Ch3LL 

I tested with 
```
# salt --version
salt 2016.11.0
# python --version
Python 2.6.6
```

and I see the same problem.

@s0undt3ch did some research on memory leaks and and specifically said that there is an issue on python 2.6 but that was long time ago and apparently was fixed https://github.com/saltstack/salt/pull/2799 .

**Comment by rallytime** [comment]
Closed via #46817
